Refactor test_flows.py (patch 6)

arnout opened the pull request:
This is hopefully the last of the test_flows refactoring.
Everything is moved down into separate classes, with a model of the test devices.
At this point, it is not yet committed to unittest or boardfarm (or something else) - it can still go either way.

test_flows.py is no longer specific for docker - except for the main function.
Since it's not clear yet in which direction we'll go further, the way this will be refactored will depend on that choice.
But the tests themselves are completely agnostic of whether they run in docker or on a device.
All the tests are still kept in a single test_flows.py file, since the way they should be split up is probably different for unittest and boardfarm.

Fixes: #896

arnout committed:
tests: environment: add Radio class

An abstract representation of a radio. For now, it is only used to store
the MAC address.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented:
Note that this is still a draft PR. The description is the dream towards which I'm working, the commits don't fully reflect that yet.

Since this PR will again conflict heavily with all other PRs affecting test flows, I will not consider it ready until #965 and #1020 are merged and this is rebased on top of that.

Galvien commented:
I would also suggest running a PEP8 formatting on top of everything.
In addition to making the file neater and more readable, any time I clean up using auto-format in vs-code a few hundred lines change and I assume I'm not the only one this happens to.
vs-code's default formatter is either Black or PEP8 so I assume just running one of these should be fine.
Some formatting tools also make sure all of the strings are kept in the same format, i.e. either double quotation or single quotation marks.

arnout committed:
tests: environment.py: add ALEntity

An ALEntity stands for a controller or agent entity. It's an abstract
class with a single implementation: ALEntityDocker.

Move the AL-MAC addresses and the UCC socket into ALEntity. The code to
derive these is moved into ALEntityDocker, so that requires just the
name of the container to be initialised.

For convenience, the UCC socket functionality (cmd_reply,
dev_get_parameter, and dev_send_1905) is accessible directly on
ALEntity.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: remove tcpdump_proc and bridge_name

These were leftovers that weren't removed in the commits removing their
use.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: refactor check_log with wait_for_log

check_log is actually a kind of wait_for_log with a really small
timeout. Refactor things so check_log calls wait_for_log with a timeout
of 300ms. This also removes the .1s sleep hack that was needed before.

Swap the order of the file reading and the timeout check. We want to
check for timeout just before sleeping, not afterwards, otherwise the
last sleep is useless.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: remove optimal_path_dummy test

Remove the optimal path dummy test because:
- It sometimes fails (cfr. #711).
- It needs a specific configuration which doesn't exist if some of the
  other tests (e.g. ap_config_bss_tear_down) have executed before.
- It makes the upcoming check_log refactoring more complicated since
  it uses the monitor log (nothing else does).
- Work is ongoing for a much improved optimal path test in UTAF.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: move wait_for_log to ALEntity and Radio classes

Instead of using the container and process names, we want to use the
ALEntity and Radio classes. Therefore, move the wait_for_log
functionality in those classes. It's an abstract function that is
implemented in the docker subclasses.

Since we need very similar functionality in two different classes, put
it in a private helper function _docker_wait_for_log.

TestFlows still has the check_log and wait_for_log wrappers, because it
has to handle failures (i.e. increment the error counter). Since the
error message is already printed by _docker_wait_for_log, we can't use
fail() or we'd have two error messages. Therefore, refactor the
non-message part of fail() into __fail_no_message() and call that.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: move beerocks_cli_command to environment module

beerocks_cli is always executed on the controller, so it can be launched
directly from the environment.

To make this possible, add a "prplmesh_command" method to AlEntity, and
add an installdir argument to its __init__. This makes abstraction of
the different install locations that you will get on different devices.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: move connection map to separate module

To clean up the test_flows module a bit, move the connection map
functionality to a separate module. This is now possible because
beerocks_cli_command can be called directly from the environment module
and no longer needs a TestFlows object.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: environment.py: add Station and VirtualAP classes

Station is just a placeholder with a MAC address. The actual handling is
done in the VirtualAP class. This will allow the same Station object to
be used for both a agent running on a device (where a real station will
be used) and an agent running in docker (where it's just dummy bwl
events that are sent).

Station has a create() method that generates a MAC address. This ensures
that everywhere we use a station, a different MAC address is used, so
different tests have less chance of interfering with each other. It does
mean that when a physical station is used, its MAC address will have to
be set explicitly before associating.

Move bwl_event_send into the docker implementation of VirtualAP, and use
its associate and disassociate methods instead. This also removes the
need for docker_command() in TestFlows.

With this, the repeater1 and repeater2 names no longer need to be stored
in TestFlows, and the init() function doesn't do much any more.
Therefore, remove it and call launch_environment_docker directly from
main.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: remove rootdir and installdir from TestFlows

installdir is no longer used. rootdir is only used for the tcpdump path.
Move that use to the main function.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented on tests/environment.py:
```diff
@@ -0,0 +1,175 @@
+import sniffer
+
+
+class AlEntity:
```
Not really PEP8 naming but probably a good idea indeed.

arnout commented on tests/environment.py:
```diff
@@ -0,0 +1,175 @@
+    This provides basic information about the entity, e.g. its AL MAC address. How this information
+    is retrieved is implementation-specific.
+
+    It also provides an abstract interface to interact with the entity, e.g. for sending CAPI
```
Running commands on the device.

arnout commented on tests/environment.py:
```diff
@@ -0,0 +1,175 @@
+agents = []
+
+
+# Concrete implementation with docker
```
I prefer to do keep it in a single file until we start doing the boardfarm stuff. I expect that to require moving stuff around between the abstract classes and the implementations, and that's easier if it's in a single file. Well, not by much, but still.

arnout commented on tests/environment.py:
```diff
@@ -0,0 +1,175 @@
+                re.search(r'ucc_listener_port=(?P<port>[0-9]+)', config_file.read()).group('port')
+
+        device_ip_output = self.command('ip', '-f', 'inet', 'addr', 'show', self.bridge_name)
+        device_ip = re.search(r'inet (?P<ip>[0-9.]+)', device_ip_output.decode('utf-8')).group('ip')
```
True, but that wasn't there before and there's already plenty in this PR :-)

arnout commented on tests/environment.py:
```diff
@@ -0,0 +1,175 @@
+        inspect_result = subprocess.run(docker_network_inspect_cmd, check=True,
+                                        stdout=subprocess.PIPE)
+
+    inspect = json.loads(inspect_result.stdout)
```
json.loads will raise an exception if it's not JSON, and the lookups will raise KeyError. So I don't see why we should bother.

arnout commented on tests/sniffer.py:
```diff
@@ -0,0 +1,46 @@
+            while not self.tcpdump_proc.poll():
+                line = self.tcpdump_proc.stderr.readline()
+                debug(line.decode()[:-1])  # strip off newline
+                if line.startswith(b"tcpdump: listening on " + self.interface.encode()):
```
Good point, but not for this PR.

VladyslavTupikin commented on tests/environment.py:
```diff
@@ -0,0 +1,299 @@
+        while True:
+            with open(logfilename) as logfile:
+                for line in logfile.readlines():
+                    if re.search(regex, line):
```
This implementation does not include a case when logfile contains 2 or more same words. It just takes the first found word.

arnout replied:
Well, that feature is something that you added, no?

rmelotte reviewed (approved):
LGTM
